### ISSUE TITLE
Feature Request for Following

### ISSUE THREAD

**Issue by deronsizemore** [InstaPy/InstaPy#247]
One strategy I came across a while back for following and getting return followers is by viewing a hashtags posts and scrolling past the "top" ones until you see "recent." Go through recent posts for that hashtag and follow all the people liking those recent posts. The idea here is that since these are recent posts that likely means the people who liked the post also did so recently. Chances are that when you follow them, there's a chance they are still actively browsing and will see your follow notification and follow you back. I've had decent success with this method and I think it would be a great feature to be able to do this with InstaPy if possible.

**Comment by sunnypatel165** [comment]
Isn't this already there?
when you say 
```
  .like_by_tags(['#cat'], amount=100)\
```
It goes and finds 100 images of cat - top ones + recent ones 
Additionally, if you specify,
```
session.set_do_follow(enabled=True, percentage=50, times=2)
```
follow 50% of those 100 liked images. 
Is this what you are requesting?
Correct me if I'm wrong @timgrossmann ?

<======================================================>
Also @timgrossmann @tibor - is there a specific order when I want to specify both like_by_tags and ste_do_follow? I just realised my following has NEVER worked! 😱 😨 

Here's my script:
```
.login()\
  .set_lower_follower_count(limit = 750) \
  .set_dont_like(dont_like) \
  .set_do_comment(True, percentage=40)\
  .set_comments(['Nice!', 'Looks really good:D'])\
  .like_by_tags(['#eatingfortheinsta' , '#foodblog', '#foodgasm'], amount=15)\
  .set_do_follow(enabled=True, percentage=50, times=2)\
  .end()
```

**Comment by deronsizemore** [comment]
I may be wrong, but I don't think so. Here's how I understand what you gave above (and I may be completely wrong)

> session.set_do_follow(enabled=True, percentage=50, times=2)

Would follow those folks who posted images with a hashtag of `#cat`, correct? It won't follow users that liked images inside of that hashtag. Also, I think if you set `percentage=50` you're not following 50% of those 100 liked images. You're following every 50th user from those images (or at least that's what it says in the examples on the front github page).

Maybe that's why it seems your following hasn't ever worked? You're finding 15 images within your specified tags and then setting the follow to be every 50th one?

**Comment by sunnypatel165** [comment]
Okay, that dawned on me now!
Apologies!

Also, I misconstrued your question! 
Yes, If you want to **"follow people who liked a recent image"** - that's a new feature and I'd vouch for that one! It does work for me too!

**Comment by timgrossmann** [comment]
@sunnypatel165 Yes, the order matters.
Always use the `.set_.....` first, since they only **set** parameters of the script.
Use the actual commands `unfollow_users`, `like_by_tag` at the end. Else they will be ignored.

**Comment by sunnypatel165** [comment]
Ahh okay, does this deserve to be in the README? (I can add)
Doesn't seem to be something that naturally strikes one while using.

**Comment by timgrossmann** [comment]
@sunnypatel165 Feel free to do so!
All the examples have this structure, but it might be best to add a more detailed explanation of the structure.

**Comment by tibor** [comment]
Is there a difference between `unfollow_users` `like_by_tag` and `like_by_tag` `unfollow_users` as well?

**Comment by timgrossmann** [comment]
@tibor Well, depends on what you mean. 
They simply get executed after each other, so this would be more "useful":

```python
.unfollow_users(...)
.like_by_tag(...)
```

Else you will instantly unfollow the with `like_by_tag` followed accounts again and they won't see you've followed them.

**Comment by tibor** [comment]
@timgrossmann Thanks, then I’ve set up my script wrong as well (that was my issue with #213)

**Comment by deronsizemore** [comment]
> Else you will instantly unfollow the with like_by_tag followed accounts again and they won't see you've followed them.

@timgrossmann Interesting... So the accounts to follow are collected from the like_by_tag command? 

So I have:

```
like_by_tags = ['#cat', '#dog', '#mouse']

.set_do_follow(True, percentage=3, times=2)
.like_by_tags(like_by_tags, amount=25)
```

That's going to grab 25 posts to like from each of my three hashtags and then follow every third account from the 75 total accounts collected?

**Comment by pekapa** [comment]
EDIT: This is wrong, see @timgrossmann comment below.

---

@deronsizemore No. This will follow every single images from those three tags. To follow one in every three images you need to set it with `percentage=0.33` (or, more clearly, `percentage=1./3.`)

You'll grab 25 images from each tag, totalling 75.

**Comment by timgrossmann** [comment]
@pekapa @deronsizemore No and no 😄

`percentage=3`means 3%
`percentage=33` is what you want <- Follows ~every third

**Comment by tibor** [comment]
BTT:

A similar feature was already discussed in #197.

Two things to consider:
1. You can’t click on the number of likes in the web interface. So you can get the user names only when there are not too many likes on a picture already.
2. InstaPy is imho not optimized for real time but for being so slow that the bot liking isn’t noticed by Instagram. For this feature we would imho have to change the focus from navigating & waiting to reloading & liking fast

**Comment by deronsizemore** [comment]
@timgrossmann Hmm, ok. I guess my confusion came from reading the docs on the front github page where it says:

```
#default enabled=False, follows ~ every 10th user from the images, times=1 (only follows a user once (if unfollowed again))

session.set_do_follow(enabled=True, percentage=10, times=2)
```
I assumed "every 10th user" came from the percentage=10 attribute?

@tibor: ah, I see. I didn't think about this using the web interface to do everything. I see what you're saying. Makes sense. Thanks

**Comment by timgrossmann** [comment]
@deronsizemore Yes, 10% means every 10th 😄 
But I'll change that to avoid confusion!

**Comment by deronsizemore** [comment]
@timgrossmann 

>Yes, 10% means every 10th 😄

Wait... so percentage=10 *does* mean every 10th or no? If it does in fact mean every 10th image, how's it arrive at that?

**Comment by sunnypatel165** [comment]
🤣 
Check the updated doc @deronsizemore for clarity!
Its 10%.. and **every 10th** is wrt that ```amount``` of images like_by_tag is 100.

**Comment by deronsizemore** [comment]
@sunnypatel165 Ah, that makes sense. Thanks

**Comment by maverjk** [comment]
is there a way to save my actual following accounts before to start any session? I would like to avoid to unfollow my actual friends.

**Comment by timgrossmann** [comment]
@maverjk Nope, you can add them to the `friends_list` like in the example, but you can't *save* your current list of friends.

**Comment by jeffreylm** [comment]
session.set_do_follow(enabled=True, percentage=10, times=2)

What is the effect of setting the "times" value to 1 or 2?

Thanks in advance.

**Comment by jeffreylm** [comment]
Tell me if the following re-write explains this function better?

#default enabled=False, follows ~ 10% of the users from the images, times=1 (only follows a user once (if unfollowed again))

session.set_do_follow(enabled=True, percentage=10, times=2)


#default enabled=False, follows ~ 10% of the users from the images _**defined/found by the settings set  for session.like_by_tags settings. This function is executed after all the hashtags have been liked and being that it is a enviromental parameter should be added previous to the session.like_by_tags so that the set parameters can be used as input to the  this fuction **_n  in the , times=1 (only follows a user once (if unfollowed again))


In short

session.set_do_follow(enabled=True, percentage=10, times=2) uses the values from session.like_by_tags to define which people to follow, but must be placed early / start of the script to work properly.


session.set_do_follow(enabled=True, percentage=10, times=2)

**Comment by pekapa** [comment]
@jeffreylm 
That's almost right.
If I'm not mistaken, it follows the user DURING the like period. Not after.
I.E. while going through the images and liking them the script randomly chooses the ones that should be liked so that, in average, the given percentage of images have their owner being followed.

Times N allows the script to follow a given user up to N times.

Basically, all the "set" functions must be placed before the "like" functions as they are responsible for getting the parameters needed to like (and comment and follow) images.

**Comment by jeffreylm** [comment]
@pekapa  Then how does the % part of the follow? So if I tell it follow 60%, it could never follow 60% of the liked accounts if during the tag & like it sometimes decided that an account is inappropriate.

For it to do it properly, the follow has to gather the liked from the database and follow the percentage of the liked. Think about it for a moment. 

I saw at times the script decided not to like an image because there was another tag with something that flagged it as inappropriate, and on top of that. It the follow did not happen, if I am not mistaken.

Am I seeing it clearly?

**Comment by jeffreylm** [comment]
@pekapa When would it be a good idea to follow someone N times? So more than once.
